### PR TITLE
Studio: Download the latest wp-cli and bundle it

### PR DESCRIPTION
**Pull request by kozer**
Fixes https://github.com/Automattic/dotcom-forge/issues/7320

## Proposed Changes

This PR do the following:
1. Adds wp-cli in the studio bundle.
2. Extract and update it when the user starts the app.

## Testing Instructions

1. Run `nvm use && npm i`
2. Notice that the WP-CLI has been downloaded in `wp-files/wp-cli` folder.
3. Run `npm run package`
4. Ensure that `wp-cli.phar` exists in `./out/<architecture>/resources` folder under `wp-files` folder.
5. Start studio app.
6. Ensure that the `wp-cli.phar` is updated and is now located under `~/.config/Studio/server-files` directory.

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: scripts/download-wp-server-files.ts [hunk 1]
```diff
@@ -17,6 +17,12 @@ const FILES_TO_DOWNLOAD = [
 		description: 'SQLite files',
 		url: 'https://downloads.wordpress.org/plugin/sqlite-database-integration.zip',
 	},
+	{
+		name: 'wp-cli',
+		description: 'WP-CLI phar file',
+		url: 'https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar',
+		destinationPath: path.join( WP_SERVER_FILES_PATH, 'wp-cli' ),
+	},
```
Comment (sejas) [inline scripts/download-wp-server-files.ts:25]: Nice!

File: src/setup-wp-server-files.ts [hunk 2]
```diff
@@ -48,9 +49,28 @@ async function copyBundledSqlite() {
...
 
+async function copyBundledWPCLI() {
+	const bundledWPCLIInstalled = await fs.pathExists( getWpCliPath() );
+	console.log( `Copying bundled WP-CLI...` );
+	console.log( bundledWPCLIInstalled );
+	if ( bundledWPCLIInstalled ) {
+		return;
+	}
+	const bundledWPCLIPath = path.join( getResourcesPath(), 'wp-files', 'wp-cli', 'wp-cli.phar' );
+	await fs.copyFile( bundledWPCLIPath, getWpCliPath() );
```
Comment (sejas) [inline src/setup-wp-server-files.ts]: Do we need to copy it?
Maybe `getWpCliPath` can return `getResourcesPath(), 'wp-files', 'wp-cli', 'wp-cli.phar'` and use it directly?

Do you know if the update function could replace the bundled version?

File: src/setup-wp-server-files.ts [hunk 3]
```diff
@@ -48,9 +49,28 @@ async function copyBundledSqlite() {
...
+}
+
+async function updateLatestWPCLI() {
+	const shouldOverwrite = true;
+	const res = await downloadWPCLI( shouldOverwrite );
+	console.log( `WP-CLI downloaded:`, res );
+}
```
Comment (sejas) [inline src/setup-wp-server-files.ts]: Let's remove updateLatestWPCLI and handle it in a different issue https://github.com/Automattic/dotcom-forge/issues/7326

I would expect that it compares the current version with the latest release and download it only if necessary

File: src/setup-wp-server-files.ts [hunk 4]
```diff
@@ -48,9 +49,28 @@ async function copyBundledSqlite() {
...
+
 export default async function setupWPServerFiles() {
 	await copyBundledLatestWPVersion();
 	await copyBundledSqlite();
+	await copyBundledWPCLI();
 	await updateLatestWordPressVersion();
 	await updateLatestSqliteVersion();
+	await updateLatestWPCLI();
```
Comment (sejas) [inline src/setup-wp-server-files.ts]: if we use the bundled version, then we don't need to copy it each time the app starts.

**Review by sejas — COMMENTED** [review]
@kozer , thank you for creating this PR.
I tested it and the bundled .phar is present in `wp-files` when running `npm install` and also in the App resources.

Screenshot 2024-05-28 at 21 11 20

Screenshot 2024-05-28 at 21 11 39

I left a comment asking if is possible to use that file directly, instead of copying it to the "home" directory.

When starting the app I also confirm the app downloads and replaces the .phar file in the home directory. But I think we don't need that behaviour. My concern is that it could make the app starting slower.

File at "~/Library/Application Support/Studio/server-files/wp-cli.phar"

Screenshot 2024-05-28 at 20 54 32

File: src/setup-wp-server-files.ts (hunk 3 above)
Comment (kozer) [reply src/setup-wp-server-files.ts]: Yes, I agree, this is a better approach! Let's do that!

File: src/setup-wp-server-files.ts (hunk 2 above)
Comment (wojtekn) [reply src/setup-wp-server-files.ts]: @sejas, would it make sense to follow the same approach for all bundled libraries, so we always store the latest version in the Studio working directory?

Comment (kozer) [reply src/setup-wp-server-files.ts]: I'm not sure if it's writable. However, let's do that, and if we want to update, and it doesn't work, then copy it into the home directory.

**Comment by kozer** [comment]
Thank you @sejas for your comments! I updated the PR! Can you test it again? Thanks!

**Review by sejas — APPROVED** [review]
Looking great. Thanks for the changes @kozer 
I tried it by running `npm run make`
I confirm the bundled version exists and I even executed inside Studio adding some code.

```
getWpCliPath(): /Users/macbookpro/Downloads/Studio.app/Contents/Resources/wp-files/wp-cli/wp-cli.phar
WP-CLI 2.10.0
```

**Comment by wojtekn** [comment]
@sejas @kozer what is the reason and benefit of handling WP CLI in a different way than WP core and SQLite plugin?

**Comment by kozer** [comment]
> @sejas @kozer what is the reason and benefit of handling WP CLI in a different way than WP core and SQLite plugin?

Hey @wojtekn ! The benefit I see here, is that, as we don't update it currently, there is no reason to add overhead when we start the app, to move thing around.
I believe thought, that when we add update into the mix, this is gonna change, as it most probably won't be able to update it as is, and we ll need to copy it into the home folder first, as implemented in my first commit under this PR.

**Comment by wojtekn** [comment]
@kozer, so why would we do this in the first place? Wouldn't it be beneficial to have it consistent for all three libraries, what I pointed out in https://github.com/Automattic/studio/pull/178#discussion_r1618413993 comment?

**Comment by kozer** [comment]
> @kozer, so why would we do this in the first place? Wouldn't it be beneficial to have it consistent for all three libraries, what I pointed out in https://github.com/Automattic/studio/pull/178#discussion_r1618413993 comment?

The benefit is only temporary, but it is valid, as is. Not a big deal though, as it will probably change soon.
If you think this is not ok, I can quickly restore that back.
What do you think?

**Comment by sejas** [comment]
we can restore the `copyBundledWPCLI` call for consistency , but let's address the update logic in a different issue https://github.com/Automattic/dotcom-forge/issues/7326.

In [this commit](https://github.com/Automattic/studio/pull/178/commits/c68ca839dedcf508d3567ec88e3fb7e05bc8b32f#diff-c1a4ebb77029d831401b5e40b14a997a07bed19b01940a7b0210dd12c33178acL63-L75) we were moving the wp-cli.phar and inmediatly downloading it from the network  every time the app started, and we need to add a more complex logic to compare wp-cli versions.

The biggest argument to keep wp-cli.phar inside the app is that its an "executable" globally available and not something attached to each site.

**Comment by kozer** [comment]
> we can restore the `copyBundledWPCLI` call for consistency , but let's address the update logic in a different issue https://github.com/Automattic/dotcom-forge/issues/7326.
> 
> In https://github.com/Automattic/studio/pull/178/commits/c68ca839dedcf508d3567ec88e3fb7e05bc8b32f#diff-c1a4ebb77029d831401b5e40b14a997a07bed19b01940a7b0210dd12c33178acL63-L75 we were moving the wp-cli.phar and inmediatly downloading it from the network every time the app started, and we need to add a more complex logic to compare wp-cli versions.
> 
> The biggest argument to keep wp-cli.phar inside the app is that its an "executable" globally available and not something attached to each site.

I started working on that. I'll handle everything in this ticket.

**Pull request merged**